research/audit_2020/attacks.py

```diff
@@ -0,0 +1,115 @@
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+"""Poisoning attack library for auditing."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+from sklearn.decomposition import PCA
+from sklearn.linear_model import LogisticRegression
+
+def make_clip_aware(trn_x, trn_y, l2_norm=10):
+  """
+  trn_x: clean training features - must be shape (n_samples, n_features)
+  trn_y: clean training labels - must be shape (n_samples, )
+
+  Returns x, y1, y2
+  x: poisoning sample
+  y1: first corresponding y value
+  y2: second corresponding y value
+  """
+  x_shape = list(trn_x.shape[1:])
+  to_image = lambda x: x.reshape([-1] + x_shape)
+  flatten = lambda x: x.reshape((x.shape[0], -1))
+  assert np.allclose(to_image(flatten(trn_x)), trn_x)
+
```

Comment:
Could you please add some general comments what the implementation of this function does?